Support credit card payments through Adyen's API

luukveenis opened the pull request:
Currently we only support payments that go through Adyen's Hosted Payment Pages (HPP). This change adds a new payment method for credit cards and everything required to support processing credit card payments directly through Adyen's API, using their client side encryption (CSE) library.

Credit card payments through Adyen work similar to HPP payments in that it's an asynchronous API. We make a payment request, Adyen tells us whether it was received successfully or not, and at some point in the future we get a notification if it succeeded or not. Most of the handling of notifications uses the existing HPP code with minimal modifications. We can capture, cancel, refund, and partially refund payments through Solidus and all operations will be handled like HPP, where the payment state is put into `processing` on a request and transitioned after we receive a notification from Adyen.

One of the challenges with Adyen is that they don't support typical payment profiles provided by services like Stripe or Braintree. Instead, Adyen supports "recurring payment contracts". These still store the user's card details with Adyen and allow them to be re-used, but they require an initial payment request to be created (such as an authorization). Because of this, we have to do the authorization during the `payment` checkout step when we still have access to the encrypted credit card data. That means by the time we get to `confirm` we've already authorized the payment and either fake the authorization to advance the payment or simply `capture` at that point, depending if the payment method has auto-capture enabled or not.

The `payment_profiles_supported?` method returns false because Solidus assumes you're following the typical flow when this is true, which causes a number of issues. However, the recurring contract is still created and the reference is stored in the `gateway_customer_profile_id` attribute on the source so that we can re-use the details to create payments or make modifications later.

The biggest issue with authorizing during the payment step is that we may be left with outstanding authorizations if the user abandons checkout on the confirm step. This is something to be aware of and anyone using credit card payments should periodically cancel these open authorizations.

mamhoff committed:
Embryonic Credit Card Gateway

mamhoff committed:
Add basic gateway methods

We have a provider class and a way to reference the preferences.

mamhoff committed:
Add special form with encrypted fields

The form holds JS include tags for custom JS as well as
JS include tags for Adyens Easy Encryption library - to be
fetched from their servers using the CSE token.

mamhoff committed:
Make Adyen Credit Card payment method available

mamhoff committed:
Include Adyen JS on the payment form

mamhoff committed:
Remove Postgres from the gemspec

We can safely assume the host app has a Database driver.

mamhoff committed:
Enable validations on CC form

mamhoff committed:
Add Adyen Credit Card Gateway factory

mamhoff committed:
Add Feature test with JS for testing validations

mamhoff committed:
Validate Encrypted Card form on submit - Step 1

This has been surprisingly hard. Spree adds submit handlers to the form,
and we do as well. They conflict with each other, so I have to unbind and re-bind them.

mamhoff committed:
Use CreditCard#encrypted_data attribute

mamhoff committed:
Add IDs to encrypted fields

These are useful for testing and still do not submit data to the server.

mamhoff committed:
Form validation feature spec

mamhoff committed:
Remove version constraints

mamhoff committed:
First working CSE payment

mamhoff committed:
Correct spec description

mamhoff committed:
Extract api_username into method

mamhoff committed:
Extract api_password method

mamhoff committed:
Order configuration methods alphabetically

mamhoff committed:
Spec out successful authorize run

mamhoff committed:
Document and test payment_profiles_supported?

mamhoff committed:
Test AdyenCreditCard#create_profile

mamhoff committed:
Add purchase method to CC gateway

mamhoff committed:
Add Unit test for gateway#purchase

mamhoff committed:
Implementation of payment modifying actions

mamhoff committed:
Lock Webmock to 1.x, add cassettes

The webmock gem has been updated to a new version which does not work
with the VCR gem yet. Locking it to version 1.x enables recording episodes!

mamhoff committed:
Fix shared mock context to work with any class under test

mamhoff committed:
Fix test suite, filter sensitive data

mamhoff committed:
Extract common gateway methods to module

This is beneficial for sharing code as well as for unifying behaviour.
The existing HPP Gateway's behaviour has preference.

mamhoff committed:
Use extracted gateway options module for HPP, too

mamhoff committed:
Make specs run without API credentials

We want full integration specs to run, but we don't want actual API
credentials to be present for that.

In order to accomplish this, I needed to download the JS library Adyen offers
and use the downloaded version in the specs.

mamhoff committed:
Add unhappy path to integration spec

mamhoff committed:
Adapt unit tests to passing the order number as reference

mamhoff committed:
Make test settings not leak into production

mamhoff committed:
Make CC payments backend viewable

mamhoff committed:
User user id OR order number as reference

For a non-logged in user, use the order number (he won't be able to re-use his
credit card). For a logged-in user, use the user id: She'll be able to re-use
her credit card this way.

mamhoff committed:
I18nize the error message when cc data is invalid

Commit:
Add optional source parameter

This gets called from adyen_payment.rb for HPP payments, where it
doesn't receive a source param. However, it also gets called from
refund.rb in Solidus Core where it does receive the source.

Commit:
Don't unbind submit events when selecting Adyen

This can cause issues in stores that have other event handlers bound to
the submit event on the form. Since Spree.disableSaveOnClick attaches an
anonymous function as an event handler, we can't simply unbind it.
Instead, we just re-enable the button if the Adyen validations fail and
the submission is cancelled.

Commit:
Support common base controller for adyen

This lets us add customizations to a single place, rather than having to
target multiple controllers.

Commit:
Use refusal reason as fallback for error messages

Commit:
Use the correct card to update payments

I ran into an issue where simply getting the last card returned from
Adyen was not the new one we just created. This caused the payment to be
updated with incorrect details. Since the new card is created earlier in
this method, we know for sure it will be the latest one created.

Commit:
Use order number as merchant reference

The order_id set in gateway_options is actually the order number and the
payment number joined with a dash.
  ie: R123456-A34HG4
Since we are using the merchant reference as the foreign key to
associate orders with notifications, this will never associate
notifications with credit cards.

Commit:
Ignore duplicate notifications from Adyen

Adyen has an "at-least-once" delivery system for notifications, meaning
that we may receive the same notification multiple times. Instead of
trying to create the notification again and rescuing from the exception,
simply check if it exists and respond with accept if it does.

From the Adyen docs:
"Duplicate notifications have the same corresponding values for the
eventCode and pspReference fields."

Commit:
Use poltergeist as JS driver

Commit:
Add ability to require factories outside of gem

Commit:
Add checks to payment flow for Adyen credit cards

The end goal is to treat all Adyen payments (HPP or credit cards) the
same during the checkout flow. We can't do that immediately because
notifications aren't associated with credit card payments correctly, so
when a payment is transitioned to `processing`, it never gets
transitioned out of it because it doesn't receive the success
notification. When this is done, we should stop returning false.

Commit:
Fix potential error in conditional

Adyen credit cards simply use Spree::CreditCard as the source, which
does not provide a #requires_manual_refund? method. To avoid hacking
this into Spree::CreditCard check if it responds to the method, since
credit cards never require a manual refund.

Commit:
Add spec for payment method check

Commit:
Refactor credit card payments

This includes a number of major changes to how credit cards work in the
gem:
- notifications are now associated with credit card payments
- credit card payments are transitioned based on notifications like HPP
- support for payment profiles was removed
- authorisation happens in the payment step
- auto_capture is properly supported on the cc payment method

The reason to auth in the payment step is that Adyen did not like that
we were doing $0 auths to "create profiles". Since this is the only step
we have access to the encrypted card data, we have to perform the auth
here. The only alternatives are to change the checkout flow or persist
the encrypted card data (neither of which are desirable).

Payment profile support was removed because Adyen doesn't truly support
payment profiles. They support the notion of recurring payment
contracts, which require a first payment transaction to store the
customer's details. This is why there was initially a $0 authorization
that passed in the flag to create a recurring contract.

I had to override #purchase! for credit cards to only call #capture!
instead, since we'll have already authorized the payment at that point.

NOTE: The biggest concern with this implementation is that if a customer
abandons checkout in the confirm step, there will be an outstanding
authorization on the card. Anyone using this functionality should
regularly cancel old authorizations.

Commit:
Create recurring contracts for Adyen credit cards

This is largely the logic that happened in #create_profile before. We
still want to create a recurring contract with Adyen so that we store
the user's credit card info in case we need to use it to modify the
payment later.

Although payment_profiles_supported is still false for Adyen credit
cards, default Solidus will still list the cards on the payment step to
allow users to select them. This maintains support for that by calling
authorise_recurring_payment using the recurring contract ID instead when
we do the authorisation.

Commit:
Fix failing feature specs

Since the authorization is happening in a different step, I mostly just
needed to re-record the VCR cassettes. I also updated the test card data
used in the specs to be valid since the old ones had expired.

Commit:
Add feature spec for refunding adyen credit cards

Testing this is a little weird due to the async notifications. This runs
an order through checkout to get a valid adyen payment, but then sends
it a fake capture notification to complete the payment. After that it
refunds it through the admin.

cbrunsdon commented:
I worked closely with luuk (and before him martin) and am :+1: on merging this. It should be completely inconsequential to anyone not using CC's.

Some bits of it are pretty unfortunate, but what part of trying to connect adyen to solidus isn't.

Dkendal commented:
requires a rebase

Dkendal commented on app/controllers/spree/adyen_notifications_controller.rb:
```diff
@@ -29,5 +30,12 @@ def authenticate
     end
+
+    def notification_exists? params
+      AdyenNotification.where(
```
`.exists?` would also be appropriate.

luukveenis committed:
Log error response for failed credit card auths

If the initial call to authorize an Adyen credit card payment fails, we
should create a log entry for it. This provides better record keeping
and also allows us to display errors where we need them.

luukveenis committed:
Provide billing address with CC payment requests

The billing address parameter is optional when making API requests with
Adyen, but since it's used for AVS checks it should be provided by
default.

The Adyen gem currently does not provide a way to pass in the billing
address, so this required some overrides to implement. I'm going turn
this into a PR and send it up to Adyen. If they approve it, we should
remove the overrides here.

Dkendal commented on solidus-adyen.gemspec:
```diff
@@ -19,14 +19,12 @@ Gem::Specification.new do |spec|
 
   spec.add_runtime_dependency "adyen", "~> 2.2.0"
-  spec.add_runtime_dependency "solidus_core", "~> 1.1"
+  spec.add_runtime_dependency "solidus_core"
```
any reason for unlocking the dep?

Dkendal commented:
thanks guys lgtm

oleksiidrobiazko commented:
@luukveenis and other guys, thank you, great work.
Could you tell me please it's not released? In master but not released, right?

Dkendal commented:
@oleksiidrobiazko you are correct, this is currently an unreleased change in master
